### PR TITLE
web: PWA improvements

### PR DESCRIPTION
**Pull request by Vadim-Karpenko**
As discussed in [https://github.com/renpy/renpy/pull/4152](https://github.com/renpy/renpy/pull/4152), some devs may require to update the `pwa_catalog.json` without re-building the game. This pull request adds the script to the web folder which does the following:
- Find current `pwa_catalog.json`
- Go through the game folder and get a list of all files
- Generate MD5 hashes and save them back to the catalog

### CODE REVIEW

**Comment by Vadim-Karpenko** [comment]
In addition, I'm including pwa_catalog.json in web3 ([c3d46b6](https://github.com/renpy/renpy-build/pull/54/commits/c3d46b6b8066522de553dbfe211aa7298731228f)). Please do not merge this pull request until https://github.com/renpy/renpy/pull/4152 is merged.

**Review by qinlili23333 — APPROVED** [review]

File: runtime/web/shell.html [hunk 1]
```diff
@@ -158,7 +158,6 @@
     <progress id="statusProgress" value="0" max="100"></progress>
   </div>
 
-  <script src="renpy-pre.js"></script>
   <script>
     // PWA
     window.addEventListener('load', function() {
```
Comment (qinlili23333) [inline runtime/web/shell.html:163]: don't need to do after load. Installing service worker is async function, without await it won't stuck page loading process. Just remove `addEventListener` and directlty register it.
and don't need to register every time. Register a registered service worker will trigger update check of service worker, this should only use when manually update is needed.
I would like to use this below.
```
if (navigator.serviceWorker && !navigator.serviceWorker.controller) {
    navigator.serviceWorker.register('./service-worker.js', { scope: '/' });
}
```

**Review by qinlili23333 — CHANGES_REQUESTED** [review]

File: runtime/web/service-worker.js [hunk 2]
```diff
@@ -10,51 +10,64 @@ var filesToCache = [
...
+      var url = filesToCache[i];
+      var cached_response = await caches.match(url);
+      var stripped_url = url.replace("/", "")
+      if (stripped_url == "") {
+        // If url is root, set it to index.html
+        stripped_url = "index.html"
+      }
+      if (cached_response) {
+        // Make sure the file is present in pwa_catalog.json
+        if (pwa_catalog["core"][stripped_url] !== undefined) {
```
[the remark this answers is not in the document]
Comment (Vadim-Karpenko) [reply runtime/web/service-worker.js:41]: Right now it is only core files in the cache, so it is not really necessary. But feel free to create a pull request to my repo, I'll merge it

**Comment by renpytom** [comment]
I don't think we want to support updating the PWA catalog outside of rebuilding the game. It seems error prone, since some files will be outside and others inside the web app.

File: runtime/web/shell.html [hunk 3]
```diff
@@ -160,11 +160,14 @@
 
   <script>
     // PWA
-    if (navigator.serviceWorker && !navigator.serviceWorker.controller) {
+    if (navigator.serviceWorker) {
+      if (navigator.serviceWorker.controller) {
```
Comment (qinlili23333) [inline runtime/web/shell.html:164]: navigator.serviceWorker.controller means a registered service worker, so you missed a `!`, only when no controller should do register

**Comment by qinlili23333** [comment]
> I don't think we want to support updating the PWA catalog outside of rebuilding the game. It seems error prone, since some files will be outside and others inside the web app.

Sometimes a developer needs to change just a few files in resource, such as simply modify a cg picture. In these circumstances, it's quicker to just replace the file in release and rebuild catalog, comparing to use RenPy launcher to release it again.
But yes we should carefully process those files outside the web app.

File: runtime/web/shell.html (hunk 3 above)
Comment (Vadim-Karpenko) [reply runtime/web/shell.html]: Thank you for pointing this, I uploaded an update to fix it

**Comment by renpytom** [comment]
I really don't want the external script - it's something extra that needs to be maintained, that duplicates the preferred way to do something I'm going to remove it so we don't need to support it.

**Pull request merged**